libkatana_python_native/src/Reductions.cpp

```diff
@@ -0,0 +1,95 @@
+struct ForGAccumulator {
+  template <typename T>
+  using type = katana::GAccumulator<T>;
+};
+
+struct ForReduceMin {
+  template <typename T>
+  using type = katana::GReduceMin<T>;
+};
+
+struct ForReduceMax {
+  template <typename T>
+  using type = katana::GReduceMax<T>;
+};
+
+// The following two structs allow GReduceLogicalOr/And to be passed to
+// ReducibleFunctor despite not being templated.
+struct ForReduceLogicalOr {
+  template <typename T>
+  using type = katana::GReduceLogicalOr;
+};
+
+struct ForReduceLogicalAnd {
+  template <typename T>
+  using type = katana::GReduceLogicalAnd;
+};
+
+template <typename For>
+struct ReducibleFunctor {
+  template <typename T>
+  py::object operator()(py::module_& m, const char* name) {
+    py::class_<typename For::template type<T>> cls(
+        m, name,
+        "A reducer object that can updated with new values and combines the "
+        "results efficiently using the appropriate operator.\n"
+        "\n"
+        "This class can be passed into numba compiled code and it's methods "
+        "can be used from there.\n");
+    cls.def(py::init<>());
+    cls.def(
+        py::init([](T v) {
+          auto self = std::make_unique<typename For::template type<T>>();
+          self->update(v);
+          return self;
+        }),
+        "Create a new instance with an initial value.");
+    katana::RegisterNumbaClass(cls);
+    katana::DefWithNumba<const T&>(
+        cls, "update", &For::template type<T>::update,
+        "Update this reducer with ``v`` performing the operation.");
+    katana::DefWithNumba<>(
+        cls, "reduce", &For::template type<T>::reduce,
+        "Get the current value of the reducer. This must only be called from "
+        "single threaded code.");
+    katana::DefWithNumba<>(
+        cls, "get_local", &For::template type<T>::getLocal,
+        "Get a sub-result of the reducers operation. This is generally the "
+        "reduced value for this thread.");
+    katana::DefWithNumba<>(
+        cls, "reset", &For::template type<T>::reset,
+        "Reset the reducer to its zero. This must only be called from single "
+        "threaded code.");
+    katana::DefConventions(cls);
+    return std::move(cls);
+  }
+};
+
+}  // namespace
+
+void
+katana::python::InitReductions(py::module_& m) {
+  katana::InstantiateForStandardTypes(
+      m, "ReduceSum", ReducibleFunctor<ForGAccumulator>());
+  katana::InstantiateForStandardTypes(
+      m, "ReduceMax", ReducibleFunctor<ForReduceMax>());
+  katana::InstantiateForStandardTypes(
+      m, "ReduceMin", ReducibleFunctor<ForReduceMin>());
+  // These calls are complex because they would usually be made from the
+  // template above. However, it's better than duplicating the implementation of
+  // ReducibleFunctor.
+  ReducibleFunctor<ForReduceLogicalOr>().operator()<bool>(m, "ReduceOr");
+  ReducibleFunctor<ForReduceLogicalAnd>().operator()<bool>(m, "ReduceAnd");
```

Comment:
And here, we are defining the class name for ReduceOr with specific metadata class metadata, as well as some of the conventional methods. 

Why exactly are we using an operator here? I don't see the benefit of using this compared to another templated function.

Reply:
`operator()` *is* a template function. :-) The reason I have to explicitly reference `operator()` instead of just using call syntax is that C++ doesn't allow `ReducibleFunctor<ForReduceLogicalOr>()<bool>(m, "ReduceOr")`. I don't totally understand the reasons why, but it boils down to the fact that call operator expansion (going from `f()` to `f.operator()()`) doesn't allow for template arguments. If fact checking again it appears it doesn't allow it EVER. So I will rename `operator()` to something else since not everyone knows that operators are just normal methods with special syntax associated with them and a few rules.

Reply:
That's interesting. I did not know that C++ disallowed that - seems strange that this work around is even permitted if the other is not.

Reply:
Yep. There are a lot of cases where C++ allows a work around, but does not allow doing something directly. I find it maddening.